prometheus/desc.go

```diff
@@ -161,6 +163,62 @@ func NewDesc(fqName, help string, variableLabels []string, constLabels Labels) *
 	return d
 }
 
+// NewConstrainedDesc allocates and initializes a new Desc with constraints on
+// the variable labels. Errors are recorded in the Desc and will be reported on
+// registration time. variableLabels, constLabels and constrainedLabels can be
+// nil if no such labels should be set. fqName must not be empty.
+//
+// variableLabels only contain the label names. Their label values are variable
+// and therefore not part of the Desc. (They are managed within the Metric.)
+//
+// constrainedLabels contains the label names and functions to normalize them.
+// Label names in constrainedLabels are merged into variableLabels, so they can
+// be omitted from variableLabels. Order of variableLabels is preserved, and new
+// labels from constrainedLabels are appended to variableLabels.
+//
+// For constLabels, the label values are constant. Therefore, they are fully
+// specified in the Desc. See the Collector example for a usage pattern.
+func NewConstrainedDesc(
+	fqName, help string,
+	variableLabels []string,
+	constLabels Labels,
+	constrainedLabels ConstrainedLabels,
+) *Desc {
+	if len(constrainedLabels) > 0 {
+		// Merge constrainedLabels into variableLabels as a source for variable
+		// labels in the Desc
+		labelSet := map[string]bool{}
+		for label := range constrainedLabels {
+			labelSet[label] = true
+		}
+		for _, label := range variableLabels {
+			labelSet[label] = false
+		}
+		for label := range constrainedLabels {
+			if labelSet[label] {
+				variableLabels = append(variableLabels, label)
+			}
+		}
+	}
+
+	d := NewDesc(fqName, help, variableLabels, constLabels)
+	if len(constrainedLabels) > 0 {
+		d.labelConstraints = make([]func(string) string, len(d.variableLabels))
+		for i, label := range d.variableLabels {
+			if fn, ok := constrainedLabels[label]; ok {
+				d.labelConstraints[i] = fn
+			} else {
+				d.labelConstraints[i] = identityConstraint
+			}
+		}
+	}
+	return d
+}
+
+func identityConstraint(lv string) string {
```

Comment:
What does it mean?

Reply:
Do we need this? Why not just leaving non constrained labels alone - not do anything special, just use old flow. Then only constrained will execute function

Reply:
This is an identity function that always returns the string it receives. This reason we need this is in the [comment above](https://github.com/prometheus/client_golang/pull/1151#discussion_r1003624180)

Reply:
proposed idea that perhaps removes this.